File: numbers.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity, get_jwt
from datetime import datetime
from src.models.user import User, db
from src.models.phone_number import PhoneNumber
from src.models.sms_provider import SMSProvider

numbers_bp = Blueprint('numbers', __name__)

@numbers_bp.route('/numbers', methods=['GET'])
def get_numbers():
    """Get available phone numbers with optional filtering"""
    try:
        # Get query parameters
        country_code = request.args.get('country_code')
        status = request.args.get('status', 'available')
        limit = min(int(request.args.get('limit', 20)), 100)  # Max 100
        offset = int(request.args.get('offset', 0))
        
        # Build query
        query = PhoneNumber.query
        
        if status:
            query = query.filter_by(status=status)
        
        if country_code:
            query = query.filter_by(country_code=country_code.upper())
        
        # Get total count
        total = query.count()
        
        # Get paginated results
        numbers = query.offset(offset).limit(limit).all()
        
        return jsonify({
            'numbers': [number.to_dict() for number in numbers],
            'total': total,
            'limit': limit,
            'offset': offset
        }), 200
        
    except Exception as e:
        return jsonify({
            'error': 'Internal Server Error',
            'message': 'An error occurred while fetching numbers',
            'code': 'FETCH_ERROR'
        }), 500

@numbers_bp.route('/numbers', methods=['POST'])
@jwt_required()
def assign_number():
    """Assign an available phone number to the authenticated user"""
    try:
        current_user_id = get_jwt_identity()
        data = request.get_json()
        
        # Validate required fields
        if not data.get('phone_number_id'):
            return jsonify({
                'error': 'Validation Error',
                'message': 'phone_number_id is required',
                'code': 'MISSING_FIELD'
            }), 400
        
        phone_number_id = data['phone_number_id']
        duration_hours = data.get('duration_hours', 1)
        
        # Validate duration
        if duration_hours < 1 or duration_hours > 24:
            return jsonify({
                'error': 'Validation Error',
                'message': 'Duration must be between 1 and 24 hours',
                'code': 'INVALID_DURATION'
            }), 400
        
        # Get the phone number
        phone_number = PhoneNumber.query.get(phone_number_id)
        if not phone_number:
            return jsonify({
                'error': 'Not Found',
                'message': 'Phone number not found',
                'code': 'NUMBER_NOT_FOUND'
            }), 404
        
        # Check if number is available
        if phone_number.status != 'available':
            return jsonify({
                'error': 'Conflict',
                'message': 'Phone number is not available',
                'code': 'NUMBER_NOT_AVAILABLE'
            }), 409
        
        # Check if user already has an active number
        existing_assignment = PhoneNumber.query.filter_by(
            user_id=current_user_id,
            status='assigned'
        ).first()
        
        if existing_assignment and not existing_assignment.is_expired():
            return jsonify({
                'error': 'Conflict',
                'message': 'You already have an active phone number assignment',
                'code': 'ALREADY_ASSIGNED',
                'details': {
                    'existing_number': existing_assignment.to_dict()
                }
            }), 409
        
        # Assign the number
        phone_number.assign_to_user(current_user_id, duration_hours)
        
        return jsonify(phone_number.to_dict()), 201
        
    except Exception as e:
        return jsonify({
            'error': 'Internal Server Error',
            'message': 'An error occurred while assigning number',
            'code': 'ASSIGNMENT_ERROR'
        }), 500

@numbers_bp.route('/numbers/<int:number_id>', methods=['GET'])
@jwt_required()
def get_number(number_id):
    """Get details of a specific phone number"""
    try:
        current_user_id = get_jwt_identity()
        
        phone_number = PhoneNumber.query.get(number_id)
        if not phone_number:
            return jsonify({
                'error': 'Not Found',
                'message': 'Phone number not found',
                'code': 'NUMBER_NOT_FOUND'
            }), 404
        
        # Check if user has access to this number
        if phone_number.user_id != current_user_id and phone_number.status != 'available':
            return jsonify({
                'error': 'Forbidden',
                'message': 'Access denied to this phone number',
                'code': 'ACCESS_DENIED'
            }), 403
        
        return jsonify(phone_number.to_dict()), 200
        
    except Exception as e:
        return jsonify({
            'error': 'Internal Server Error',
            'message': 'An error occurred while fetching number details',
            'code': 'FETCH_ERROR'
        }), 500

@numbers_bp.route('/numbers/<int:number_id>', methods=['DELETE'])
@jwt_required()
def release_number(number_id):
    """Release an assigned phone number back to available pool"""
    try:
        current_user_id = get_jwt_identity()
        
        phone_number = PhoneNumber.query.get(number_id)
        if not phone_number:
            return jsonify({
                'error': 'Not Found',
                'message': 'Phone number not found',
                'code': 'NUMBER_NOT_FOUND'
            }), 404
        
        # Check if user owns this number
        if phone_number.user_id != current_user_id:
            return jsonify({
                'error': 'Forbidden',
                'message': 'You are not authorized to release this number',
                'code': 'NOT_AUTHORIZED'
            }), 403
        
        # Release the number
        phone_number.release()
        
        return jsonify({
            'message': 'Phone number released successfully'
        }), 200
        
    except Exception as e:
        return jsonify({
            'error': 'Internal Server Error',
            'message': 'An error occurred while releasing number',
            'code': 'RELEASE_ERROR'
        }), 500

@numbers_bp.route('/numbers/my', methods=['GET'])
@jwt_required()
def get_my_numbers():
    """Get phone numbers assigned to the current user"""
    try:
        current_user_id = get_jwt_identity()
        
        # Get user's assigned numbers
        numbers = PhoneNumber.query.filter_by(user_id=current_user_id).all()
        
        # Clean up expired numbers
        for number in numbers:
            if number.is_expired() and number.status == 'assigned':
                number.release()
        
        # Refresh the query after cleanup
        numbers = PhoneNumber.query.filter_by(user_id=current_user_id).all()
        
        return jsonify({
            'numbers': [number.to_dict() for number in numbers]
        }), 200
        
    except Exception as e:
        return jsonify({
            'error': 'Internal Server Error',
            'message': 'An error occurred while fetching your numbers',
            'code': 'FETCH_ERROR'
        }), 500

@numbers_bp.route('/numbers/<int:number_id>/extend', methods=['POST'])
@jwt_required()
def extend_number(number_id):
    """Extend the assignment duration of a phone number"""
    try:
        current_user_id = get_jwt_identity()
        data = request.get_json()
        
        additional_hours = data.get('additional_hours', 1)
        
        # Validate additional hours
        if additional_hours < 1 or additional_hours > 24:
            return jsonify({
                'error': 'Validation Error',
                'message': 'Additional hours must be between 1 and 24',
                'code': 'INVALID_DURATION'
            }), 400
        
        phone_number = PhoneNumber.query.get(number_id)
        if not phone_number:
            return jsonify({
                'error': 'Not Found',
                'message': 'Phone number not found',
                'code': 'NUMBER_NOT_FOUND'
            }), 404
        
        # Check if user owns this number
        if phone_number.user_id != current_user_id:
            return jsonify({
                'error': 'Forbidden',
                'message': 'You are not authorized to extend this number',
                'code': 'NOT_AUTHORIZED'
            }), 403
        
        # Check if number is still assigned
        if phone_number.status != 'assigned':
            return jsonify({
                'error': 'Conflict',
                'message': 'Phone number is not currently assigned',
                'code': 'NOT_ASSIGNED'
            }), 409
        
        # Extend the assignment
        if phone_number.expires_at:
            # If not expired, extend from current expiry time
            if not phone_number.is_expired():
                phone_number.expires_at = phone_number.expires_at + timedelta(hours=additional_hours)
            else:
                # If expired, extend from current time
                phone_number.expires_at = datetime.utcnow() + timedelta(hours=additional_hours)
        else:
            # If no expiry set, set from current time
            phone_number.expires_at = datetime.utcnow() + timedelta(hours=additional_hours)
        
        db.session.commit()
        
        return jsonify({
            'message': 'Phone number assignment extended successfully',
            'number': phone_number.to_dict()
        }), 200
        
    except Exception as e:
        return jsonify({
            'error': 'Internal Server Error',
            'message': 'An error occurred while extending number assignment',
            'code': 'EXTEND_ERROR'
        }), 500

File: numbers.js
const express = require('express');
const Joi = require('joi');
const PhoneNumber = require('../models/PhoneNumber');
const { requireAdmin } = require('../middleware/auth');

const router = express.Router();

// Validation schemas
const assignNumberSchema = Joi.object({
  phoneNumber: Joi.string().pattern(/^\+[1-9]\d{1,14}$/).optional(),
  countryCode: Joi.string().length(2).uppercase().optional(),
  duration: Joi.number().min(1).max(24).default(1) // hours
});

const extendAssignmentSchema = Joi.object({
  duration: Joi.number().min(1).max(24).required() // additional hours
});

// @route   GET /api/numbers/available
// @desc    Get available phone numbers
// @access  Private
router.get('/available', async (req, res, next) => {
  try {
    const { countryCode, limit = 10, provider } = req.query;

    // Validate query parameters
    const querySchema = Joi.object({
      countryCode: Joi.string().length(2).uppercase().optional(),
      limit: Joi.number().min(1).max(50).default(10),
      provider: Joi.string().valid('twilio', 'nexmo', 'vonage', 'other').optional()
    });

    const { error, value } = querySchema.validate({ countryCode, limit: parseInt(limit), provider });
    if (error) {
      return res.status(400).json({
        success: false,
        error: {
          code: 'VALIDATION_ERROR',
          message: 'Invalid query parameters',
          details: error.details.map(detail => ({
            field: detail.path[0],
            message: detail.message
          }))
        }
      });
    }

    // Build query
    const query = { status: 'available' };
    if (value.countryCode) query.countryCode = value.countryCode;
    if (value.provider) query.provider = value.provider;

    const numbers = await PhoneNumber.find(query)
      .limit(value.limit)
      .select('phoneNumber countryCode provider cost currency capabilities')
      .sort({ createdAt: -1 });

    const total = await PhoneNumber.countDocuments(query);

    res.json({
      success: true,
      data: {
        numbers: numbers.map(number => ({
          ...number.toJSON(),
          formattedNumber: number.formattedNumber
        })),
        total,
        limit: value.limit,
        available: numbers.length
      }
    });

  } catch (error) {
    next(error);
  }
});

// @route   POST /api/numbers/assign
// @desc    Assign a phone number to user
// @access  Private
router.post('/assign', async (req, res, next) => {
  try {
    const { error, value } = assignNumberSchema.validate(req.body);
    if (error) {
      return res.status(400).json({
        success: false,
        error: {
          code: 'VALIDATION_ERROR',
          message: 'Validation failed',
          details: error.details.map(detail => ({
            field: detail.path[0],
            message: detail.message
          }))
        }
      });
    }

    const { phoneNumber, countryCode, duration } = value;
    const userId = req.user._id;

    // Check if user already has active numbers (limit to prevent abuse)
    const activeNumbers = await PhoneNumber.countDocuments({
      user: userId,
      status: 'assigned'
    });

    const maxActiveNumbers = req.user.role === 'admin' ? 50 : 5;
    if (activeNumbers >= maxActiveNumbers) {
      return res.status(400).json({
        success: false,
        error: {
          code: 'LIMIT_EXCEEDED',
          message: `Maximum ${maxActiveNumbers} active numbers allowed`
        }
      });
    }

    let numberToAssign;

    if (phoneNumber) {
      // Assign specific number
      numberToAssign = await PhoneNumber.findOne({
        phoneNumber,
        status: 'available'
      });

      if (!numberToAssign) {
        return res.status(404).json({
          success: false,
          error: {
            code: 'NUMBER_NOT_AVAILABLE',
            message: 'Phone number is not available'
          }
        });
      }
    } else {
      // Find any available number
      const query = { status: 'available' };
      if (countryCode) query.countryCode = countryCode;

      numberToAssign = await PhoneNumber.findOne(query);

      if (!numberToAssign) {
        return res.status(404).json({
          success: false,
          error: {
            code: 'NO_NUMBERS_AVAILABLE',
            message: 'No phone numbers available'
          }
        });
      }
    }

    // Assign number to user
    await numberToAssign.assignToUser(userId, duration);

    // Populate user info
    await numberToAssign.populate('user', 'firstName lastName email');

    res.json({
      success: true,
      data: {
        assignment: {
          id: numberToAssign._id,
          phoneNumber: numberToAssign.phoneNumber,
          formattedNumber: numberToAssign.formattedNumber,
          countryCode: numberToAssign.countryCode,
          provider: numberToAssign.provider,
          status: numberToAssign.status,
          assignedAt: numberToAssign.assignedAt,
          expiresAt: numberToAssign.expiresAt,
          user: numberToAssign.user
        }
      },
      message: 'Phone number assigned successfully'
    });

  } catch (error) {
    next(error);
  }
});

// @route   GET /api/numbers/my
// @desc    Get user's assigned phone numbers
// @access  Private
router.get('/my', async (req, res, next) => {
  try {
    const { status, page = 1, limit = 20 } = req.query;
    const userId = req.user._id;

    // Validate query parameters
    const querySchema = Joi.object({
      status: Joi.string().valid('assigned', 'expired', 'released').optional(),
      page: Joi.number().min(1).default(1),
      limit: Joi.number().min(1).max(100).default(20)
    });

    const { error, value } = querySchema.validate({
      status,
      page: parseInt(page),
      limit: parseInt(limit)
    });

    if (error) {
      return res.status(400).json({
        success: false,
        error: {
          code: 'VALIDATION_ERROR',
          message: 'Invalid query parameters'
        }
      });
    }

    // Build query
    const query = { user: userId };
    if (value.status) {
      if (value.status === 'expired') {
        query.status = 'assigned';
        query.expiresAt = { $lt: new Date() };
      } else {
        query.status = value.status;
      }
    }

    const skip = (value.page - 1) * value.limit;

    const [numbers, total] = await Promise.all([
      PhoneNumber.find(query)
        .populate('user', 'firstName lastName email')
        .sort({ assignedAt: -1 })
        .skip(skip)
        .limit(value.limit),
      PhoneNumber.countDocuments(query)
    ]);

    // Add message count for each number
    const numbersWithStats = await Promise.all(
      numbers.map(async (number) => {
        const messageCount = await number.getMessageCount();
        return {
          ...number.toJSON(),
          formattedNumber: number.formattedNumber,
          messageCount,
          isExpired: number.isExpired()
        };
      })
    );

    res.json({
      success: true,
      data: {
        assignments: numbersWithStats,
        pagination: {
          total,
          page: value.page,
          limit: value.limit,
          pages: Math.ceil(total / value.limit)
        }
      }
    });

  } catch (error) {
    next(error);
  }
});

// @route   POST /api/numbers/:id/extend
// @desc    Extend phone number assignment
// @access  Private
router.post('/:id/extend', async (req, res, next) => {
  try {
    const { error, value } = extendAssignmentSchema.validate(req.body);
    if (error) {
      return res.status(400).json({
        success: false,
        error: {
          code: 'VALIDATION_ERROR',
          message: 'Validation failed',
          details: error.details.map(detail => ({
            field: detail.path[0],
            message: detail.message
          }))
        }
      });
    }

    const { duration } = value;
    const numberId = req.params.id;
    const userId = req.user._id;

    // Find number
    const phoneNumber = await PhoneNumber.findOne({
      _id: numberId,
      user: userId,
      status: 'assigned'
    });

    if (!phoneNumber) {
      return res.status(404).json({
        success: false,
        error: {
          code: 'NUMBER_NOT_FOUND',
          message: 'Phone number not found or not assigned to you'
        }
      });
    }

    // Extend assignment
    await phoneNumber.extendAssignment(duration);

    res.json({
      success: true,
      data: {
        assignment: {
          id: phoneNumber._id,
          phoneNumber: phoneNumber.phoneNumber,
          expiresAt: phoneNumber.expiresAt,
          extendedDuration: duration
        }
      },
      message: 'Assignment extended successfully'
    });

  } catch (error) {
    next(error);
  }
});

// @route   DELETE /api/numbers/:id
// @desc    Release phone number
// @access  Private
router.delete('/:id', async (req, res, next) => {
  try {
    const numberId = req.params.id;
    const userId = req.user._id;

    // Find number
    const phoneNumber = await PhoneNumber.findOne({
      _id: numberId,
      user: userId
    });

    if (!phoneNumber) {
      return res.status(404).json({
        success: false,
        error: {
          code: 'NUMBER_NOT_FOUND',
          message: 'Phone number not found or not assigned to you'
        }
      });
    }

    // Release number
    await phoneNumber.release();

    res.json({
      success: true,
      data: {
        assignmentId: numberId,
        releasedAt: phoneNumber.releasedAt
      },
      message: 'Phone number released successfully'
    });

  } catch (error) {
    next(error);
  }
});

// @route   GET /api/numbers/stats
// @desc    Get phone number statistics
// @access  Private
router.get('/stats', async (req, res, next) => {
  try {
    const userId = req.user._id;

    const [
      totalNumbers,
      activeNumbers,
      expiredNumbers,
      availableNumbers
    ] = await Promise.all([
      PhoneNumber.countDocuments({ user: userId }),
      PhoneNumber.countDocuments({ user: userId, status: 'assigned' }),
      PhoneNumber.countDocuments({
        user: userId,
        status: 'assigned',
        expiresAt: { $lt: new Date() }
      }),
      PhoneNumber.countDocuments({ status: 'available' })
    ]);

    res.json({
      success: true,
      data: {
        stats: {
          totalNumbers,
          activeNumbers,
          expiredNumbers,
          availableNumbers
        }
      }
    });

  } catch (error) {
    next(error);
  }
});

// Admin routes

// @route   POST /api/numbers
// @desc    Add new phone number (Admin only)
// @access  Private (Admin)
router.post('/', requireAdmin, async (req, res, next) => {
  try {
    const createNumberSchema = Joi.object({
      phoneNumber: Joi.string().pattern(/^\+[1-9]\d{1,14}$/).required(),
      countryCode: Joi.string().length(2).uppercase().required(),
      provider: Joi.string().valid('twilio', 'nexmo', 'vonage', 'other').required(),
      providerId: Joi.string().optional(),
      cost: Joi.number().min(0).default(0),
      currency: Joi.string().length(3).uppercase().default('USD'),
      capabilities: Joi.array().items(Joi.string().valid('sms', 'voice', 'mms')).default(['sms'])
    });

    const { error, value } = createNumberSchema.validate(req.body);
    if (error) {
      return res.status(400).json({
        success: false,
        error: {
          code: 'VALIDATION_ERROR',
          message: 'Validation failed',
          details: error.details.map(detail => ({
            field: detail.path[0],
            message: detail.message
          }))
        }
      });
    }

    const phoneNumber = new PhoneNumber(value);
    await phoneNumber.save();

    res.status(201).json({
      success: true,
      data: {
        phoneNumber: {
          ...phoneNumber.toJSON(),
          formattedNumber: phoneNumber.formattedNumber
        }
      },
      message: 'Phone number added successfully'
    });

  } catch (error) {
    next(error);
  }
});

// @route   GET /api/numbers/cleanup
// @desc    Cleanup expired numbers (Admin only)
// @access  Private (Admin)
router.post('/cleanup', requireAdmin, async (req, res, next) => {
  try {
    const cleanedCount = await PhoneNumber.cleanupExpired();

    res.json({
      success: true,
      data: {
        cleanedCount
      },
      message: `${cleanedCount} expired numbers cleaned up`
    });

  } catch (error) {
    next(error);
  }
});

module.exports = router;

